Introduce new parameter 'persistenceInfoLocation'

timomeinen opened the pull request:
- Introduce new parameter 'persistenceInfoLocation' to define the location of the persistence.xml file. This is needed to create a WAR archive without the persistence.xml. Spring for example is capable of setting up the persistenceUnit without persistence.xml. The value of this parameter is simply forwarded to the StaticWeaver of EclipseLink.

- Defined the provider in persistence.xml. Otherwise JEE container like Wildfly will fall back to their default Hibernate.

- Repaired the 'prefix' parameter and renamed it to 'basePackage'

ethlo commented:
Hi, and thanks for your pull-request. I do have some questions though:

> Introduce new parameter 'persistenceInfoLocation' to define the location of the persistence.xml file. This is needed to create a WAR archive without the persistence.xml. Spring for example is capable of setting up the persistenceUnit without persistence.xml. The value of this parameter is simply forwarded to the StaticWeaver of EclipseLink.

If the persistence.xml is not required, why do you need to have a way of specifying it?

> Defined the provider in persistence.xml. Otherwise JEE container like Wildfly will fall back to their default Hibernate.

Sure, you can also provide your own persistence.xml file, and it will just replace the classes referenced

> Repaired the 'prefix' parameter and renamed it to 'basePackage'

What is this for?

timomeinen commented:
> If the persistence.xml is not required, why do you need to have a way of specifying it?

The persistence.xml is required by the `StaticWeaveProcessor` class as stated [here](https://wiki.eclipse.org/EclipseLink/UserGuide/JPA/Advanced_JPA_Development/Performance/Weaving/Static_Weaving):

> There must be a META-INF directory that contains the persistence.xml file.

`StaticWeaveProcessor` weaves only classes defined in `persistence.xml` and even worse, the file has to be in the correct location. To be concrete: the weaver expects the file under `target/classes/META-INF/persistence.xml` and weaves only classes defined within the file.

In our setup, we don't want to use any persistence.xml as Spring creates the PersistenceInfo during runtime. We need the persistence.xml for the static weaving, but we don't want the file to be part of the generated war-file. One solution would be to exclude the file within the `maven-war-plugin`, but then we still have the file under `target/classes`, which is picked up when launching the application via the IDE. The solution is to define `persistenceInfoLocation` and let the plugin generate a persistence.xml for the static weaving compiler in a temporary directory in `/target` which is ignored by the IDE and the `maven-war-plugin`.

> Defined the provider in persistence.xml.

The idea was that when classes had been processed by the Eclipselink static weave compiler, it willl definitely run with Eclipselink as the JPA provider. Therefore, we could define the `provider` in the file. If deploying the war-file including the generated `persistence.xml` in a standard JEE container, like e.g. Wildfly, the default provider will be used. In this case Hibernate. This is an alternative solution to the previous one mentioned, if you want to use the generated file.

> Repaired the 'prefix' parameter and renamed it to 'basePackage'

The `prefix` parameter was defined by the old Javadoc style. I converted it to the annotation based approach. Otherwise, I was not able to use it.
If you use a framework like Spring Data, the generated persistence.xml will contain classes from Spring Data annotated with `@Converter`. These classes have optional dependencies and I don't want to use them in my project. Therefore, I need the prefix parameter to let only classes of my base-package be in the generated `persistence.xml`.
I renamed it to `basePackage`, because this term is so widely used within the Spring community (Classpath Scanning, Spring Boot), so that the other developers in my team know what this parameter means.

ethlo commented:
Thank you for your contribution! I have included it in the release of version 2.6.4, available here https://github.com/ethlo/eclipselink-maven-plugin/releases/tag/v2.6.4
